### ISSUE TITLE
Improve CountryRuleFactory

### ISSUE THREAD

**Issue by otbutz** [graphhopper/graphhopper#2404]
Currently the association between Country and CountryRule is hardcoded in the CountryRuleFactory with both available rules carrying a static instance of themselves like e.g

```java
public final static AustriaCountryRule RULE = new AustriaCountryRule();
```

This is fine for a few countries but gets clunky if we add coverage for more. I'd propose that we switch to using ServiceLoader instead and let the CountryRules tell us which country they are valid for.

```java
public class AustriaCountryRule implements CountryRule {
    
    @Override
    public Country getCountry() {
        return Country.AUT;
    }
```

The resulting factory code would be a lot more dynamic:

```java
public class CountryRuleFactory {
    
    private final Map<Country, CountryRule> rules;
    
    public CountryRuleFactory() {
        Map<Country, CountryRule> rulesMap = new EnumMap<>(Country.class);
        ServiceLoader<CountryRule> loader = ServiceLoader.load(CountryRule.class);
        for (Iterator<CountryRule> iter = loader.iterator(); iter.hasNext(); ) {
            CountryRule rule = iter.next();
            if (rule.getCountry() == null) {
                throw new IllegalStateException("CountryRule needs to define a countrycode.");
            }
            rulesMap.put(rule.getCountry(), rule);
        }
        this.rules = Collections.unmodifiableMap(rulesMap);
    }

    public CountryRule getCountryRule(Country country) {
        return rules.get(country);
    }
}
```

**Comment by karussell** [comment]
> This is fine for a few countries but gets clunky if we add coverage for more.

Why? We should not make this more complicated than necessary, especially not if it is no problem at the moment. I see a CountryRule as unique and if someone wants to improve it one can contribute or fork GH.

(I even think at the moment a fork is not necessary as you can create your own `country` encoded value.)

**Comment by otbutz** [comment]
If we only start with european countries, we'd have maintain a +40 cases switch statement and add the static `RULE` Boilercode to every implementation. (which is also prone to nasty copy&paste mistakes)

**Comment by karussell** [comment]
I do not see a big problem in a long switch statement at the moment. And the plain text file approach you proposed is much more prone to copy&paste mistakes whereas the switch statement is compile time checked.

**Comment by easbar** [comment]
I don't think we even need this switch statement. We could just create a HashMap with all countries and use this instead. Ok, in this case adding a country is a matter of creating the file *and* adding it to the HashMap, but at least this way we version-control which countries are enabled and this probably doesn't change that often.

**Comment by otbutz** [comment]
Should i craft a PR?

**Comment by easbar** [comment]
For the `HashMap` approach? Sure, if you think this helps. Maybe we can even use `EnumMap` somehow?

**Comment by easbar** [comment]
.. but is it really more concise than the switch?

```java
        switch (country) {
            case DEU: return GermanyCountryRule.RULE;
            case AUT: return AustriaCountryRule.RULE;
       }
```

vs.

```java
    map.put(DEU, GermanyCountryRule.RULE);
    map.put(AUT, AustriaCountryRule.RULE);
```

not sure tbh.

**Comment by otbutz** [comment]
Yes, because we can avoid the nasty `RULE` :wink: 

```java
rules.put(AUT, new AustriaCountryRule());
```

**Comment by otbutz** [comment]
I would also like to move the country code association into the CountryRule as a rule should know its own countrycode IMHO.

```java
public class AustriaCountryRule implements CountryRule {
    
    @Override
    public Country getCountry() {
        return Country.AUT;
    }
```

We'd only have to extend the list and the risk of copy&paste mistakes is near zero :)

```java
public class CountryRuleFactory {
    
    private final Map<Country, CountryRule> rules = new EnumMap<>(Country.class);
    
    public CountryRuleFactory() {
        for (CountryRule rule : getAvailableRules()) {
            rules.put(rule.getCountry(), rule);
        }
    }

    public CountryRule getCountryRule(Country country) {
        return rules.get(country);
    }
    
    private static List<CountryRule> getAvailableRules() {
        return Arrays.asList(
                        new AustriaCountryRule(),
                        new GermanyCountryRule());
    }
}
```

**Comment by easbar** [comment]
> as a rule should know its own countrycode IMHO

I don't think I agree. Country rules are just country rules. The decision which country rule is applied for which country is something different. For example what if I want to map `DEU` to some other rule, like:

```java
map.put(DEU, new MyOwnGermanyCountryRule());
```

Is there any downside to this? I mean, why do you want to move the country code association into the rules?

**Comment by otbutz** [comment]
I have a GraphHopper fork with rules for all european countries. Once you pass the 20+ country mark, copy and paste errors are easily missed:

```java
        switch (country) {
        case AUT:
            return AustriaCountryRule.RULE;
        case BEL:
            return BelgiumCountryRule.RULE;
        case BGR:
            return BulgariaCountryRule.RULE;
        case CHE:
            return SwitzerlandCountryRule.RULE;
        case CZE:
            return CzechiaCountryRule.RULE;
        case DEU:
            return GermanyCountryRule.RULE;
        case DNK:
            return DenmarkCountryRule.RULE;
        case ESP:
            return SpainCountryRule.RULE;
        case FRA:
            return FranceCountryRule.RULE;
        case HRV:
            return CroatiaCountryRule.RULE;
        case HUN:
            return HungaryCountryRule.RULE;
        case ITA:
            return ItalyCountryRule.RULE;
        case LUX:
            return LuxembourgCountryRule.RULE;
        case NLD:
            return NetherlandsCountryRule.RULE;
        case POL:
            return PolandCountryRule.RULE;
        case PRT:
            return PortugalCountryRule.RULE;
        case SRB:
            return SerbiaCountryRule.RULE;
        case SVK:
            return SlovakiaCountryRule.RULE;
        case SVN:
            return SloveniaCountryRule.RULE;
        case SWE:
            return SwedenCountryRule.RULE;
        // ....
```

**Comment by otbutz** [comment]
> ```java
> map.put(DEU, new MyOwnGermanyCountryRule());
> ```

There's nothing wrong with multiple rules sharing the same countrycode in order to override our default rules.
To make things easily pluggable for lib users:

```java
    private final Map<Country, CountryRule> rules = new EnumMap<>(Country.class);
    
    public CountryRuleFactory() {
        for (CountryRule rule : getDefaultRules()) {
            rules.put(rule.getCountry(), rule);
        }
        for (CountryRule rule : getCustomRules()) {
            rules.put(rule.getCountry(), rule);
        }
    }

    public CountryRule getCountryRule(Country country) {
        return rules.get(country);
    }
    
    protected List<CountryRule> getCustomRules() {
        return Collections.emptyList();
    }
    
    private List<CountryRule> getDefaultRules() {
        return Arrays.asList(
                        new AustriaCountryRule(),
                        new GermanyCountryRule());
    }
```

**Comment by easbar** [comment]
Hm, how about this: `CountryRule` manages a `Map<Country, CountryRule>`. It contains the default/upstream country rules, but can be modified in user code. I don't think it is necessary to differentiate 'default' and 'custom' rules. There is just a map of rules that can be modified.

```java
class CountryRuleFactory {
    private Map<Country, CountryRule> rules;
    
    CountryRuleFactory default() {
       CountryRuleFactory result = new CountryRuleFactory();
       result.put(Country.AUT, new AustriaCountryRule());
       // ... all the default rules here
    }
    
    // use this if you want to change the rules. We can also add code so we are able to remove rules, clear the entire map, or set a new map, actually we can just expose the entire map if you will
    public putRule(Country country, CountryRule rule) {
        rules.put(country, rule);
    }
    
    @Override
    CountryRule getRule(Country country) {
         return rules.get(country);
    }
}
```

Regarding the Country association I think if this is something you want to do this can be done in user code like, e.g. thi s(?):

```java
interface MyCountryRule extends CountryRule {
    Country getCountry();
}

class MyAustriaCountryRule extends MyCountryRule {
   @Override
   Country getCountry() {
       return AUT;
   }
   
   /// my country rule implementation
}
CountryRuleFactory factory = new CountryRuleFactory();
replaceCountryRule(factory, new MyAustriaCountryRule());
/// ...

void replaceCountryRule(ConutryRuleFactory factory, MyCountryRule rule) {
    factory.put(rule.getCountry(), rule.getRule());
}
```

**Comment by otbutz** [comment]
How about this:

```java
public void register(Country country, CountryRule rule)

public void remove(Country country)
```

**Comment by easbar** [comment]
I would not mind 
```java
Map<Country, CountryRule> getCountryToRuleMap();
```
Also passing a map via the constructor would be useful. I think there is no reason to hide the map somehow.

**Comment by otbutz** [comment]
Fixed by #2418
